### ISSUE TITLE
IPVS cleanup not done if SIGHUP occurs

### ISSUE THREAD

**Issue by jckasper** [acassen/keepalived#1561]
**Describe the bug**
When using the IPVS feature of keepalived, the virtual server and real servers are normally cleaned up when the keepalived service ends.  If a SIGHUP signal has arrived since the keepalived service was started, then clean up of the virtual server and real servers is not done when the keepalived service ends.

**To Reproduce**
- Configure keepalived service with 1 or more virtual server and real servers.  
- Send a SIGHUP to the service to force a reconfiguration / reload of the keepalived.conf.  
- End the keepalived service. 

**Expected behavior**
It is expected that the virtual and real servers would always get cleaned up when the keepalived service ends.

**Keepalived version**
Keepalived v2.0.19 (11/24,2019), git commit v20191114-652-g038556d644

**Distro (please complete the following information):**
 - Name: Alpine
 - Version: 3.11
 - Architecture: x86_64

**Details of any containerisation or hosted service (e.g. AWS)**
If keepalived is being run in a container on a hosted service, provide full details

**Configuration file:**
```
global_defs {
    vrrp_mcast_group4 224.0.0.18
}

vrrp_instance vip-192.68.94.10 {
    state BACKUP
    interface eth1
    virtual_router_id 105
    priority 100
    nopreempt
    virtual_ipaddress {
        192.68.94.10
    }
}

virtual_server 192.68.94.10 1884 {
  delay_loop 5
  lvs_sched rr
  lvs_method TUN
  protocol TCP

  real_server 10.188.230.81 31797 {
    HTTP_GET {
      connect_port 31904
      url {
        path /
        status_code 200
      }
    }
  }

  real_server 10.188.230.121 31797 {
    HTTP_GET {
      connect_port 31904
      url {
        path /
        status_code 200
      }
    }
  }
}
```

**Notify and track scripts**
None

**System Log entries**
Keepalived service termination (when no SIGHUP):

```
2020/04/22 13:50:53.180434 keepalived | Stopping
2020/04/22 13:50:53.180584 keepalived | (vip-192.68.94.10) sent 0 priority
2020/04/22 13:50:53.180658 keepalived | (vip-192.68.94.10) removing VIPs.
2020/04/22 13:50:53.185124 keepalived | Shutting down service [10.188.230.121]:tcp:1884 from VS [192.68.94.10]:tcp:1884
2020/04/22 13:50:53.185393 keepalived | Shutting down service [10.188.230.81]:tcp:1884 from VS [192.68.94.10]:tcp:1884
2020/04/22 13:50:53.188090 keepalived | Stopped - used 0.000000 user time, 0.088859 system time
2020/04/22 13:50:54.211635 keepalived | Stopped - used 0.016424 user time, 0.065666 system time
2020/04/22 13:50:54.213236 keepalived | Stopped Keepalived v2.0.19 (11/24,2019), git commit v20191114-652-g038556d644
2020/04/22 13:50:54.213510 /usr/sbin/keepalived has ended
```

Keepalived service termination (after SIGHUP occurred earlier):

```
2020/04/22 14:04:37.283908 keepalived | Stopping
2020/04/22 14:04:37.284178 keepalived | (vip-192.68.94.10) sent 0 priority
2020/04/22 14:04:37.284185 keepalived | (vip-192.68.94.10) removing VIPs.
2020/04/22 14:04:38.325271 keepalived | Stopped - used 0.063159 user time, 0.170022 system time
2020/04/22 14:04:38.326600 keepalived | Stopped Keepalived v2.0.19 (11/24,2019), git commit v20191114-652-g038556d644
2020/04/22 14:04:38.327679 /usr/sbin/keepalived has ended
```

**Did keepalived coredump?**
No

**Additional context**
None

**Comment by pqarmitage** [comment]
As a quick workaround, you could try using `lvs_flush_onstop VS` or `lvs_flush_on_stop` if there is no other IPVS configuration on the system.

I will have a look to try and find the cause of the problem.

**Comment by jckasper** [comment]
We looked into that, but there are multiple IPVS configurations on the node so that was not an option.   I did scan through the cleanup and SIGHUP re-config code some, but I did not see anything obvious.

**Comment by pqarmitage** [comment]
My recollection is that using `ipvs_flush_on_stop VS` should be safe since it is designed to only remove the virtual servers that keepalived has set up. It would be interesting to know if you still get the problem using that option.

**Comment by jckasper** [comment]
A couple items:
1. I did not even know that there was a `VS` flag on the `lvs_flush_onstop` option.  I was looking here: https://www.keepalived.org/manpage.html and this optional `VS` is not listed.  I think the man page should be updated so that users are aware that this flag is available. 

2. I tried out the `lvs_flush_onstop VS` option and I got mixed results:  
- There were a few times in which stopping the keepalived service after the SIGHUP resulted in the virtual server/real servers being left around.  
- There were also a few times in which the clean up of the virtual/real server IPVS rules actually happened.

The difference seemed to be if data was received on the virtual server AFTER the SIGHUP, then everything cleaned up fine.  If there was no data received to the virtual server AFTER the SIGHUP, then the virtual server and real server IPVS rules were not cleaned up.  I am not sure if this was conclusive or just a coincidence. 

There are 2 child thread/processes: Healthcheck and VRRP.  On the success case I see:

```
2020/04/28 20:56:24.847068 keepalived | Stopped - used 0.010364 user time, 0.031139 system time
2020/04/28 20:56:25.877302 keepalived | Stopped - used 0.018736 user time, 0.029989 system time
2020/04/28 20:56:25.879643 keepalived | Stopped Keepalived v2.0.19 (11/24,2019), git commit v20191114-652-g038556d644
```

But on the failure case (when IPVS rules were not cleaned up), there is only 1 log entry for `Stopped - used 0.010364 user time, 0.031139 system time`.  It seems as if one of the child processes is not being stopped.

**Comment by pqarmitage** [comment]
@jckasper I will ask Alexandre to update the man page, but if you install the keepalived-doc package, one of the man pages installed is keepalived.conf(5).

I have tried running your configuration on Alpine 3.11 with keepalived v2.0.19 and cannot reproduce the problem. Could you please provide the following additional information:

1. The commands you are using to start, reload and  stop keepalived
2.  The output of `ps -ef | grep keepalived` when keepalived is running
3. The full keepalived logs from when keepalived start to when it stops when the problem occurs.
4. If you can send a SIGUSR1 signal to the keepalived parent process before the shutdown, keepalived will create /tmp/keepalived-check.data. Copies of those files both when keepalived removes the IPVS configuration and also when keepalived doesn't would also be useful.

I have noticed that when there isn't the second`Stopped` log entry, the following log messages are also missing:
```
Shutting down service [10.188.230.121]:tcp:1884 from VS [192.68.94.10]:tcp:1884
Shutting down service [10.188.230.81]:tcp:1884 from VS [192.68.94.10]:tcp:1884
```
This looks as though the keepalived checker process is terminating abnormally while it is shutting down, before it removes the IPVS configuration.

**Comment by libesz** [comment]
@pqarmitage Most probably this is not related, but FYI official alpine build has a patch to force the pid file directory usage in keepalived: https://github.com/alpinelinux/aports/blob/master/community/keepalived/always-use-pid-dir.patch
Can this push keepalived on other code paths during SIGHUP handling?

**Comment by pqarmitage** [comment]
@libesz I was actually using the Alpine keepalived v2.0.19 package, exactly the same as you were, so the use-pid-fid patch issue will be the same for both of us.

**Comment by jckasper** [comment]
The Alpine keepalived v2.0.19 package is running inside of an IBM Cloud Kubernetes Service (IKS) cluster.  The problem has been created on IKS clusters of various sizes (3 node to 100+ nodes) and releases (1.13 - 1.17).   In an IKS cluster, the Kubernetes Load Balancer service is provided by a pair of Kubernetes pods each running the Alpine v2.0.19 package/service.  Each of the pods is defined as the BACKUP.   Kubernetes anti-affinity rules are used to force each pod to be placed on a different Kubernetes worker node.  Custom GO logic is used in the pods to auto-generate the keepalived.conf based on the current configuration of the cluster.   This custom GO logic is also responsible for performing on the fly re-configuration of the keepalived.conf file as the cluster changes (new nodes added or removed).  If the keepalived service ever stops or ends, the pod is deleted and Kubernetes will start a new pod on another worker node in the cluster. 

List of processes running in the Kubernetes pod:
```
/ # ps -ef
PID   USER     TIME  COMMAND
    1 root      0:00 /usr/local/bin/keepalived
   23 root      0:00 /usr/sbin/keepalived --dont-fork --dump-conf --log-console --log-detail --release-vips --address-monitoring
   24 root      0:00 /usr/sbin/keepalived --dont-fork --dump-conf --log-console --log-detail --release-vips --address-monitoring
   25 root      0:00 /usr/sbin/keepalived --dont-fork --dump-conf --log-console --log-detail --release-vips --address-monitoring
   27 root      0:00 sh
   32 root      0:00 ps -ef
/ #
```
PID1 = the custom GO logic
PID23, 24, 25 = keepalived service 

Starting keepalived service: The custom GO logic starts up the keepalived service with the following options (also shown above):
```
/usr/sbin/keepalived --dont-fork --dump-conf --log-console --log-detail --release-vips --address-monitoring
```

Reload of keepalived service:  The custom GO logic sends a SIGHUP to the main keepalived service process (23 in the above example).  

Stopping keepalived service:  The keepalived service is not stopped until the Kubernetes pod is deleted.  The custom GO logic catches the SIGTEM signal that Kubernetes sends to pid=1 of the Docker image and forwards the SIGTERM signal to the main keepalived process (23 in the above example). 

The full keepalived logs from start to end when the problem occurs will be difficult because the custom GO logic as well as all 3 of the keepalived processes write logs to the console.  These logs often get intermixed.  The current options that we specify also result in quite a bit of noise.  I’ll see if I can disable some of those keepalived options and recreate a cleaner / smaller log file of this issue.

As far as passing in SIGUSR1 signal and generating the /tmp/keepalived-check.data, yes I should be able to do that for both the success case (no SIGHUP) and a failure case (after SIGHUP).

**Comment by jckasper** [comment]
Ok.  I recreated the issue and I have captured the `keepalived-check.data` and the entire logs from the keepalived service from start to end.  I am attaching 3 sets of keepalived.log / keepalived_check.data:
- failure = SIGHUP was done, cleanup of the IPVS rules did NOT occur
- nosighup = SIGHUP was not done, cleanup of the IPVS rules worked fine
- worked = SIGHUP was done -AND- cleanup of the IPVS rules did occur !!!

All 3 sets of data/logs were done on the same Kubernetes worker node with the same cluster configuration.   The actual `keepalived.conf` that was used is included in each log file.  I looked through the logs and tried to find anything different between the failure / worked cases, but I didn't see anything - some operations appeared in the log in a different order, but I didn't see anything else.

[failure-keepalived_check.data.txt](https://github.com/acassen/keepalived/files/4555663/failure-keepalived_check.data.txt)
[failure-keepalived.log.txt](https://github.com/acassen/keepalived/files/4555664/failure-keepalived.log.txt)

[nosighup-keepalived_check.data.txt](https://github.com/acassen/keepalived/files/4555665/nosighup-keepalived_check.data.txt)
[nosighup-keepalived.log.txt](https://github.com/acassen/keepalived/files/4555666/nosighup-keepalived.log.txt)

[worked-keepalived_check.data.txt](https://github.com/acassen/keepalived/files/4555667/worked-keepalived_check.data.txt)
[worked-keepalived.log.txt](https://github.com/acassen/keepalived/files/4555668/worked-keepalived.log.txt)

**Comment by pqarmitage** [comment]
In the failure situation, the following log messages are missing:
```
2020/04/30 01:07:55.284887 keepalived | Shutting down service [127.0.0.11]:tcp:80 from VS [10.188.132.205]:tcp:80
2020/04/30 01:07:55.284981 keepalived | Shutting down service [10.188.230.116]:tcp:80 from VS [10.188.132.205]:tcp:80
2020/04/30 01:07:55.285162 keepalived | Shutting down service [10.188.230.121]:tcp:80 from VS [10.188.132.205]:tcp:80
2020/04/30 01:07:55.285492 keepalived | Stopped - used 0.000000 user time, 0.062740 system time
```
This is consistent with the IPVS configuration not being cleaned up, in other words when the above log messages are output the IPVS configuration is cleaned up, and when the above messages are not output then the IPVS configuration is not cleaned up.

So far as I can see from the code in sigend() in keepalived/core/main.c, a message will be logged unless a child process exits with exit code 0, and this suggests that the IPVS child process is exiting with exit code 0. For diagnostic purposes it might be worth logging in report_child_status()  when a child process exists with a 0 exit code.

@jckasper It is very difficult for me to investigate this further on my own when I cannot reproduce the problem. If I produce some diagnostic patches would you be able to test them?

**Comment by jckasper** [comment]
I agree the current log files did not show much difference between working/failing cases. 
Yes, I should be able test out some diagnostic patches.

**Comment by pqarmitage** [comment]
I have just been doing some other testing with reloading keepalived and am finding strange behaviour with the IPVS code and it is segfaulting. Certainly on occasions keepalived is not clearing the IPVS configuration, just as you are experiencing, so I am hoping it has the same cause and I can track it down.

One thing that I am seeing is that whenever the keepalived IPVS process segfaults, the parent logs details of the segfault, whereas your logs don't show that. Could you try sending SIGSEGV to the IPVS/checker/healthchecker process and see if the logs contain anything from the parent process reporting the segfault.

**Comment by jckasper** [comment]
When I have the following processes running:
```
PID   USER     TIME  COMMAND
    1 root      0:00 /usr/local/bin/keepalived
   24 root      0:00 /usr/sbin/keepalived --dont-fork --dump-conf --log-console
   25 root      0:00 /usr/sbin/keepalived --dont-fork --dump-conf --log-console
   26 root      0:00 /usr/sbin/keepalived --dont-fork --dump-conf --log-console
```
PID 1 = custom GO code
PID 24 = main keepalived service 
PID 25 = Healthcheck child process
PID 26 = VRRP child process

Problem can be re-created by sending SIGHUP to the 25 / Healthcheck child process, followed by sending SIGTEM to 24 / main keepalived service.  The SIGHUP is not terminating the Healthcheck child process (pid is still active after the SIGHUP), but it does seem to be causing some setting to occur such that the later SIGTEM does not complete cleanly.

I tried your suggestion in sending SIGSEGV to the  25 / Healthcheck child process and I got the following:

```
2020/05/05 18:25:03.558861 keepalived | Keepalived_healthcheckers exited due to segmentation fault (SIGSEGV).
2020/05/05 18:25:03.558881 keepalived |   Please report a bug at https://github.com/acassen/keepalived/issues
2020/05/05 18:25:03.558886 keepalived |   and include this log from when keepalived started, a description
2020/05/05 18:25:03.558891 keepalived |   of what happened before the crash, your configuration file and the details below.
2020/05/05 18:25:03.558896 keepalived |   Also provide the output of keepalived -v, what Linux distro and version
2020/05/05 18:25:03.558900 keepalived |   you are running on, and whether keepalived is being run in a container or VM.
...
```

If the health check process is segfaulting during the processing of SIGTERM, there is nothing being dumped in the console/logs to show that it is occurring.

**Comment by jckasper** [comment]
**Update:**  I am now able to re-create a segmentation fault (SIGSEGV) in the Healthcheck child process fairly reliably by first sending a SIGHUP to that PID and then 10 to 15 seconds later, a SIGTERM to that same PID.   This re-creates the original issue 4 out of 5 times.  The first thing that appears in the console/logs (while processing the SIGTERM) is the segmentation fault.  If the SIGHUP is not first sent to the Healthcheck child process, then the SIGTERM cleans up everything fine.  

The main keepalived service PID and the VRRP child process PID remain active during this entire time.

**Comment by pqarmitage** [comment]
Since you are getting a SIGSEGV, can you get a coredump and do a stack backtrace with gdb? If you can, it would be useful to do it several times since what I am seeing is inconsistent.

Can you also try sending the SIGHUP and SIGTERM signals to the parent keepalived process, since that is really the intended usage.

**Comment by jckasper** [comment]
The keepalived service is running in an Alpine container.  A core dump file is not being generated when the health check child process gets this segmentation fault.  How do I force the core dump  file to be generated on Alpine?

Sending SIGTERM to the main keepalived parent process terminates the Alpine / Kuberenetes pod in which the keepallived service is running... making it rather difficult to collect anything.

**Comment by pqarmitage** [comment]
@jckasper Commit d233d32 fixes the segfault I have been observing. Whilst I can't be certain that it will fix your problem, what I can say is that the problem that the commit fixes could definitely have been the cause of your problem. When the problem was occurring what I was seeing in the logs exactly matched what you were seeing.

**Comment by libesz** [comment]
@pqarmitage I mananed to collect a backtrace from a core dump from the segfault when sending SIGHUP+SIGTERM to the checker pid.
There it is:
```
Core was generated by `/usr/sbin/keepalived --dont-fork --log-console --log-detail --release-vips --ad'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  __list_del (next=0x900010008, prev=0x41) at list_head.h:114
114     list_head.h: No such file or directory.
(gdb) bt
#0  __list_del (next=0x900010008, prev=0x41) at list_head.h:114
#1  list_head_del (entry=0x55df9dd38788) at list_head.h:126
#2  thread_cancel (thread_cp=0x55df9dd38740) at scheduler.c:1427
#3  0x000055df9d0d0c86 in netlink_close (nl=0x55df9d12c3f0 <nl_kernel>) at keepalived_netlink.c:678
#4  kernel_netlink_close_monitor () at keepalived_netlink.c:2352
#5  0x000055df9d0d0cec in kernel_netlink_close () at keepalived_netlink.c:2364
#6  0x000055df9d0d2999 in checker_terminate_phase1 (schedule_next_thread=<optimized out>) at check_daemon.c:216
#7  0x000055df9d0d2a2e in start_checker_termination_thread (thread=<optimized out>) at check_daemon.c:250
#8  0x000055df9d102da4 in thread_call (thread=0x7f73de9ecd20) at scheduler.c:1820
#9  process_threads (m=0x55df9d12df20) at scheduler.c:1820
#10 0x000055df9d102e7a in launch_thread_scheduler (m=<optimized out>) at scheduler.c:1922
#11 0x000055df9d0d2fd7 in start_check_child () at check_daemon.c:672
#12 start_check_child () at check_daemon.c:555
#13 0x000055df9d0c98b8 in start_keepalived () at main.c:447
#14 keepalived_main (argc=<optimized out>, argv=<optimized out>) at main.c:2068
#15 0x00007f73dec881f5 in ?? () from /lib/ld-musl-x86_64.so.1
#16 0x00007f73dec881ce in ?? () from /lib/ld-musl-x86_64.so.1
#17 0x00007ffed9ee9f00 in ?? ()
#18 0x0000000000000000 in ?? ()
```
Hope this helps to determine if the recent fix should cover this or not. I can send the dump as well in email if needed.
I am heading to test the fix as well in the same environment.

**Comment by libesz** [comment]
I can confirm the fix eliminated the segfault and seems the services are disabled (so IPVS shall be cleaned up) properly, according to the logs:
```
# SIGHUP sent to checker PID:
2020/05/07 09:36:57.639347 keepalived | Reloading
2020/05/07 09:36:57.639478 keepalived | Got SIGHUP, reloading checker configuration
2020/05/07 09:36:57.639821 keepalived | Opening file '/etc/keepalived/keepalived.conf'.
2020/05/07 09:36:57.640452 keepalived | WARNING - virtual server [10.176.191.246]:tcp:30091 and real server [10.176.120.188]:tcp:32007 ports don't match - resetting
2020/05/07 09:36:57.640754 keepalived | WARNING - virtual server [10.176.191.246]:tcp:30091 and real server [10.176.120.183]:tcp:32007 ports don't match - resetting
2020/05/07 09:36:57.640890 keepalived | WARNING - virtual server [10.176.191.246]:tcp:30092 and real server [10.176.120.188]:tcp:30356 ports don't match - resetting
2020/05/07 09:36:57.641100 keepalived | WARNING - virtual server [10.176.191.246]:tcp:30092 and real server [10.176.120.183]:tcp:30356 ports don't match - resetting
2020/05/07 09:36:57.641165 keepalived | Reinitializing ipvs
2020/05/07 09:36:57.642257 keepalived | Activating healthchecker for service [127.0.0.11]:tcp:30091 for VS [10.176.191.246]:tcp:30091
2020/05/07 09:36:57.642543 keepalived | Activating healthchecker for service [10.176.120.188]:tcp:30091 for VS [10.176.191.246]:tcp:30091
2020/05/07 09:36:57.643158 keepalived | Activating healthchecker for service [10.176.120.183]:tcp:30091 for VS [10.176.191.246]:tcp:30091
2020/05/07 09:36:57.643702 keepalived | Activating healthchecker for service [127.0.0.11]:tcp:30092 for VS [10.176.191.246]:tcp:30092
2020/05/07 09:36:57.643863 keepalived | Activating healthchecker for service [10.176.120.188]:tcp:30092 for VS [10.176.191.246]:tcp:30092
2020/05/07 09:36:57.643982 keepalived | Activating healthchecker for service [10.176.120.183]:tcp:30092 for VS [10.176.191.246]:tcp:30092

# SIGTERM sent to checker PID
2020/05/07 09:37:05.780473 keepalived | Shutting down (inhibited) service [127.0.0.11]:tcp:30091 from VS [10.176.191.246]:tcp:30091
2020/05/07 09:37:05.780665 keepalived | Shutting down (inhibited) service [10.176.120.188]:tcp:30091 from VS [10.176.191.246]:tcp:30091
2020/05/07 09:37:05.780813 keepalived | Shutting down service [10.176.120.183]:tcp:30091 from VS [10.176.191.246]:tcp:30091
2020/05/07 09:37:05.781443 keepalived | Shutting down (inhibited) service [127.0.0.11]:tcp:30092 from VS [10.176.191.246]:tcp:30092
2020/05/07 09:37:05.781469 keepalived | Shutting down (inhibited) service [10.176.120.188]:tcp:30092 from VS [10.176.191.246]:tcp:30092
2020/05/07 09:37:05.781476 keepalived | Shutting down service [10.176.120.183]:tcp:30092 from VS [10.176.191.246]:tcp:30092
2020/05/07 09:37:05.781656 keepalived | Stopped - used 0.007861 user time, 0.031453 system time
2020/05/07 09:37:05.781974 keepalived | Healthcheck child process(37) died: Respawning
2020/05/07 09:37:05.782469 keepalived | Starting Healthcheck child process, pid=46
2020/05/07 09:37:05.782597 keepalived | Opening file '/etc/keepalived/keepalived.conf'.
2020/05/07 09:37:05.783198 keepalived | WARNING - virtual server [10.176.191.246]:tcp:30091 and real server [10.176.120.188]:tcp:32007 ports don't match - resetting
2020/05/07 09:37:05.783214 keepalived | WARNING - virtual server [10.176.191.246]:tcp:30091 and real server [10.176.120.183]:tcp:32007 ports don't match - resetting
2020/05/07 09:37:05.783220 keepalived | WARNING - virtual server [10.176.191.246]:tcp:30092 and real server [10.176.120.188]:tcp:30356 ports don't match - resetting
2020/05/07 09:37:05.783226 keepalived | WARNING - virtual server [10.176.191.246]:tcp:30092 and real server [10.176.120.183]:tcp:30356 ports don't match - resetting
2020/05/07 09:37:05.783238 keepalived | Initializing ipvs
2020/05/07 09:37:05.783246 keepalived | Registering Kernel netlink reflector
2020/05/07 09:37:05.783254 keepalived | Registering Kernel netlink command channel
```

I've also tried to manually send a SIGHUP to the checked process, than tear down and terminate the entire keepalived service, and IPVS is cleaned up.

We might wait @jckasper to confirm also the fix is effective.

**Comment by jckasper** [comment]
I tested the updated keepalived image (thanks @libesz) and the IPVS rules are now cleaned up after a SIGHUP has occurred.

**Comment by pqarmitage** [comment]
Well that's good news. Many thanks @libesz and @jckasper for your input and for testing; that was certainly a tricky issue to track down.

@jckasper are you happy to close this issue now?

**Comment by jckasper** [comment]
Yes, this issue can be closed.  Thank you for finding/fixing it.

**Comment by libesz** [comment]
@pqarmitage thanks for the fix! Do you think this may also cover https://github.com/acassen/keepalived/issues/1555 ? Since as I remember there is also a SIGHUP related misbehavior...
I can go and re-test the new file based checker configuration with the latest source tree if you think it is worth to.

**Comment by pqarmitage** [comment]
@libesz I think it is unlikely to fix #1555, but if you could test it that would be really helpful rather than me trying to chase down a problem that no longer exists.

**Comment by pqarmitage** [comment]
Commit 1344729 ensures that if a child process segfaults while terminating then the parent process will log that accordingly. One of the original problems with the issue was that there was no indication that the vrrp process had segfaulted, and indeed the parent reported that the vrrp process had successfully terminated.
